cpp/net/url_fetcher.cc

```diff
@@ -0,0 +1,200 @@
+
+  if (!req) {
+    // TODO(pphaneuf): The dreaded null request... These are fairly
+    // fatal things, like protocol parse errors, but could also be a
+    // connection timeout. I think we should do retries in this case,
+    // with a deadline of our own? At least, then, it would be easier
+    // to distinguish between an obscure error, or a more common
+    // timeout.
+    VLOG(1) << "RequestCallback received a null request";
+    state->task_->Return(Status::UNKNOWN);
+    return;
+  }
+
+  state->response_->status_code = evhttp_request_get_response_code(req);
+  if (state->response_->status_code < 100) {
+    // TODO(pphaneuf): According to my reading of libevent, this is
+    // most likely to be a connection refused?
+    VLOG(1) << "request has a status code lower than 100: "
+            << state->response_->status_code;
+    state->task_->Return(
+        Status(util::error::FAILED_PRECONDITION, "connection refused"));
+    return;
+  }
+
+  map<string, string, ci_less<string>> headers;
+  for (evkeyval* ptr = evhttp_request_get_input_headers(req)->tqh_first; ptr;
+       ptr = ptr->next.tqe_next) {
+    map<string, string>::iterator it(headers.find(ptr->key));
+
+    if (it == headers.end()) {
+      headers.insert(make_pair(ptr->key, ptr->value));
+    } else {
+      it->second.append(", ");
+      it->second.append(ptr->value);
+    }
+  }
+  state->response_->headers.swap(headers);
+
+  const size_t body_length(
+      evbuffer_get_length(evhttp_request_get_input_buffer(req)));
+  string body(reinterpret_cast<const char*>(evbuffer_pullup(
+                  evhttp_request_get_input_buffer(req), body_length)),
+              body_length);
+  state->response_->body.swap(body);
+
+  state->task_->Return();
+}
+
+
+UrlFetcher::Request AdjustRequest(UrlFetcher::Request req) {
+  // Strip the body out to save space.
+  req.body.clear();
+
+  if (req.url.Path().empty()) {
+    req.url.SetPath("/");
+  }
+
+  return req;
+}
+
+
+State::State(ConnectionPool* pool, evhttp_cmd_type verb,
+             const UrlFetcher::Request& req, UrlFetcher::Response* resp,
+             Task* task)
+    : pool_(CHECK_NOTNULL(pool)),
+      conn_(pool_->Get(req.url)),
+      request_(AdjustRequest(req)),
+      response_(CHECK_NOTNULL(resp)),
+      task_(CHECK_NOTNULL(task)),
+      http_req_(CHECK_NOTNULL(evhttp_request_new(&RequestCallback, this))) {
+  CHECK_NOTNULL(conn_.get());
+
+  if (request_.url.Protocol() != "http") {
+    VLOG(1) << "unsupported protocol: " << request_.url.Protocol();
+    task->Return(
+        Status(util::error::INVALID_ARGUMENT,
+               "UrlFetcher: unsupported protocol: " + req.url.Protocol()));
+    return;
+  }
+
+  evhttp_add_header(evhttp_request_get_output_headers(http_req_), "Host",
+                    request_.url.Host().c_str());
+
+  // Remember to use the parameter here, because the copy we keep in
+  // request_ has the body stripped out to save space.
```

Comment:
It might go away soon, I just realised that I'll need to keep the body, in case we need to retry the request. There's a way to add data to an ```evbuffer``` without it taking ownership, so I'll use that instead, and keep it in the ```std::string```.